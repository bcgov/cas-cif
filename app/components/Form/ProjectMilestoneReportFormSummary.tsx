import { useMemo } from "react";
import { graphql, useFragment } from "react-relay";
import FormBase from "./FormBase";
import readOnlyTheme from "lib/theme/ReadOnlyTheme";
import type { JSONSchema7 } from "json-schema";
import CUSTOM_DIFF_FIELDS from "lib/theme/CustomDiffFields";
import { utils } from "@rjsf/core";
import { ProjectMilestoneReportFormSummary_projectRevision$key } from "__generated__/ProjectMilestoneReportFormSummary_projectRevision.graphql";
import projectMilestoneUiSchema from "data/jsonSchemaForm/projectMilestoneUiSchema";
import { getSortedReports } from "./Functions/reportingRequirementFormChangeFunctions";
import { getMilestoneFilteredSchema } from "./Functions/getMilestoneFilteredSchema";

const { fields } = utils.getDefaultRegistry();

const customFields = { ...fields, ...CUSTOM_DIFF_FIELDS };

interface Props {
  projectRevision: ProjectMilestoneReportFormSummary_projectRevision$key;
  viewOnly?: boolean;
}

const ProjectMilestoneReportFormSummary: React.FC<Props> = (props) => {
  const { summaryMilestoneFormChanges, isFirstRevision } = useFragment(
    graphql`
      fragment ProjectMilestoneReportFormSummary_projectRevision on ProjectRevision {
        isFirstRevision
        summaryMilestoneFormChanges: formChangesFor(
          formDataTableName: "reporting_requirement"
          reportType: "Milestone"
          first: 1000
        ) @connection(key: "connection_summaryMilestoneFormChanges") {
          edges {
            # eslint-disable-next-line relay/unused-fields
            node {
              id
              # eslint-disable-next-line relay/unused-fields
              formDataRecordId
              isPristine
              newFormData
              operation
              formChangeByPreviousFormChangeId {
                newFormData
              }
              formByJsonSchemaName {
                jsonSchema
              }
            }
          }
        }
      }
    `,
    props.projectRevision
  );

  const renderDiff = !isFirstRevision && !props.viewOnly;

  // If we are showing the diff then we want to see archived records,
  // otherwise filter out the archived milestone reports
  const milestoneReportFormChanges = renderDiff
    ? summaryMilestoneFormChanges.edges
    : summaryMilestoneFormChanges.edges.filter(
        ({ node }) => node.operation !== "ARCHIVE"
      );

  // Sort consolidated milestone form change records
  const [sortedMilestoneReports] = useMemo(() => {
    return getSortedReports(milestoneReportFormChanges, true);
  }, [milestoneReportFormChanges]);

  const allFormChangesPristine = useMemo(
    () =>
      !milestoneReportFormChanges.some(
        (report) =>
          report?.node.isPristine === false || report?.node.isPristine === null
      ),
    [milestoneReportFormChanges]
  );

  const milestoneReportsJSX = useMemo(() => {
    return sortedMilestoneReports.map((milestoneReport, index) => {
      // Set the formSchema and formData based on showing the diff or not
      const reportingRequirementFormDiffObject = renderDiff
        ? getMilestoneFilteredSchema(
            milestoneReport.formByJsonSchemaName.jsonSchema
              .schema as JSONSchema7,
            milestoneReport
          )
        : {
            formSchema: milestoneReport.formByJsonSchemaName.jsonSchema.schema,
            formData: milestoneReport.newFormData,
          };

      return (
        <div key={index} className="reportContainer">
          <header>
            <h4>Milestone Report {index + 1}</h4>
          </header>
          {/* Show this part if none of milestone report form properties have been updated */}
          {Object.keys(reportingRequirementFormDiffObject.formSchema.properties)
            .length === 0 &&
            milestoneReport.operation !== "ARCHIVE" && (
              <em>Milestone report not updated</em>
            )}

          {/* Show this part if the whole milestone report has been removed */}
          {renderDiff && milestoneReport.operation === "ARCHIVE" && (
            <em className="diffOld">Milestone Report Removed</em>
          )}
          <FormBase
            key={`form-${milestoneReport.id}`}
            tagName={"dl"}
            theme={readOnlyTheme}
            fields={renderDiff ? customFields : fields}
            schema={
              reportingRequirementFormDiffObject.formSchema as JSONSchema7
            }
            uiSchema={projectMilestoneUiSchema}
            formData={reportingRequirementFormDiffObject.formData}
            formContext={{
              operation: milestoneReport.operation,
              oldData:
                milestoneReport.formChangeByPreviousFormChangeId?.newFormData,
            }}
          />

          <style jsx>{`
            div.reportContainer {
              padding-top: 1em;
            }
          `}</style>
        </div>
      );
    });
  }, [sortedMilestoneReports, renderDiff]);

  return (
    <>
      <h3>Project Milestone Reports</h3>
      {milestoneReportFormChanges.length < 1 && props.viewOnly && (
        <dd>
          <em>No Milestone Reports</em>
        </dd>
      )}
      {(allFormChangesPristine || milestoneReportFormChanges.length < 1) &&
      !props.viewOnly ? (
        <dd>
          <em>Milestone Reports not {isFirstRevision ? "added" : "updated"}</em>
        </dd>
      ) : (
        milestoneReportsJSX
      )}
    </>
  );
};

export default ProjectMilestoneReportFormSummary;
