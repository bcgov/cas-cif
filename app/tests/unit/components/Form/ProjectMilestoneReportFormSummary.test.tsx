import { graphql } from "react-relay";
import { screen } from "@testing-library/react";
import ComponentTestingHelper from "tests/helpers/componentTestingHelper";
import compiledProjectMilestoneReportFormSummaryQuery, {
  ProjectMilestoneReportFormSummaryQuery,
} from "__generated__/ProjectMilestoneReportFormSummaryQuery.graphql";
import ProjectMilestoneReportFormSummary from "components/Form/ProjectMilestoneReportFormSummary";
import milestoneProdSchema from "../../../../../schema/data/prod/json_schema/milestone.json";

const testQuery = graphql`
  query ProjectMilestoneReportFormSummaryQuery @relay_test_operation {
    query {
      projectRevision(id: "Test Project Revision ID") {
        ...ProjectMilestoneReportFormSummary_projectRevision
      }
    }
  }
`;

const mockQueryPayload = {
  Form() {
    return {
      jsonSchema: milestoneProdSchema,
    };
  },
  ProjectRevision() {
    const result = {
      isFirstRevision: false,
      summaryMilestoneFormChanges: {
        edges: [
          {
            node: {
              id: "Test Reporting Requirement ID - 1",
              isPristine: false,
              newFormData: {
                totalEligibleExpenses: 1000,
                description: "charmander",
                projectId: 1,
                reportingRequirementIndex: 1,
                reportType: "General",
                reportDueDate: "2020-01-10T23:59:59.999-07:00",
                reportingRequirementId: 1,
                hasExpenses: true,
                calculatedNetAmount: 111,
                calculatedGrossAmount: 999,
                calculatedHoldbackAmount: 888,
                adjustedNetAmount: 11,
                adjustedGrossAmount: 99,
                adjustedHoldbackAmount: 88,
              },
              operation: "UPDATE",
              formChangeByPreviousFormChangeId: {
                newFormData: {
                  description: "bulbasaur",
                  projectId: 1,
                  reportingRequirementIndex: 1,
                  reportDueDate: "2020-01-01T13:59:59.999-07:00",
                  reportType: "Advanced",
                  reportingRequirementId: 1,
                  hasExpenses: true,
                },
              },
              formDataRecordId: 1,
            },
          },
          {
            node: {
              id: "Test Reporting Requirement ID - 2",
              isPristine: false,
              newFormData: {
                description: "Removed comment",
                projectId: 1,
                reportDueDate: "2020-01-07T23:59:59.999-07:00",
                reportingRequirementIndex: 1,
                reportingRequirementId: 2,
              },
              operation: "ARCHIVE",
              formChangeByPreviousFormChangeId: {
                newFormData: {
                  description: "Removed comment",
                  projectId: 1,
                  reportDueDate: "2020-01-05T23:59:59.999-07:00",
                  reportingRequirementIndex: 1,
                  reportingRequirementId: 2,
                },
              },
              formDataRecordId: 2,
            },
          },
        ],
      },
    };
    return result;
  },
};

const defaultComponentProps = {
  setValidatingForm: jest.fn(),
  onSubmit: jest.fn(),
};

const componentTestingHelper =
  new ComponentTestingHelper<ProjectMilestoneReportFormSummaryQuery>({
    component: ProjectMilestoneReportFormSummary,
    testQuery: testQuery,
    compiledQuery: compiledProjectMilestoneReportFormSummaryQuery,
    getPropsFromTestQuery: (data) => ({
      query: data.query,
      projectRevision: data.query.projectRevision,
    }),
    defaultQueryResolver: mockQueryPayload,
    defaultQueryVariables: {},
    defaultComponentProps: defaultComponentProps,
  });

describe("The Project Milestone Report Form Summary", () => {
  beforeEach(() => {
    jest.resetAllMocks();
    componentTestingHelper.reinit();
  });

  it("Only displays the data fields that have changed", () => {
    componentTestingHelper.loadQuery();
    componentTestingHelper.renderComponent();

    // changed fields
    expect(screen.getByText("Milestone Description")).toBeInTheDocument();
    expect(screen.getByText("Milestone Type")).toBeInTheDocument();

    // Archive milestone report
    expect(screen.getByText("Milestone Report")).toBeInTheDocument();
    const milestoneReport = document.querySelector("dd > em.diffOld");
    expect(milestoneReport.textContent).toBe("Milestone Report");
  });

  it("Displays diffs of the the data fields that were updated", () => {
    componentTestingHelper.loadQuery();
    componentTestingHelper.renderComponent();

    // description diff
    expect(screen.getByText("bulbasaur")).toBeInTheDocument();
    expect(screen.getByText("charmander")).toBeInTheDocument();

    // milestone diff
    expect(screen.getByText("General")).toBeInTheDocument();
    expect(screen.getByText("Advanced")).toBeInTheDocument();

    // report due date diff
    expect(screen.getByText(/Jan[.]? 1, 2020/i)).toBeInTheDocument();
    expect(screen.getByText(/Jan[.]? 10, 2020/i)).toBeInTheDocument();

    // total eligible expenses
    expect(
      screen.getByText("Total Eligible Expenses (optional)")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$1,000\.00/i)).toBeInTheDocument();

    // calculated values
    expect(
      screen.getByText("Gross Payment Amount This Milestone")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$999\.00/i)).toBeInTheDocument();
    expect(
      screen.getByText("Net Payment Amount This Milestone")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$888\.00/i)).toBeInTheDocument();
    expect(
      screen.getByText("Holdback Amount This Milestone")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$111\.00/i)).toBeInTheDocument();

    // adjusted values
    expect(
      screen.getByText("Gross Payment Amount This Milestone (Adjusted)")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$99\.00/i)).toBeInTheDocument();
    expect(
      screen.getByText("Net Payment Amount This Milestone (Adjusted)")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$11\.00/i)).toBeInTheDocument();
    expect(
      screen.getByText("Holdback Amount This Milestone (Adjusted)")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$88\.00/i)).toBeInTheDocument();
  });

  const latestCommittedData = {
    latestCommittedMilestoneFormChanges: {
      edges: [
        {
          node: {
            newFormData: {
              totalEligibleExpenses: 1000,
              description: "charmander",
              projectId: 1,
              reportingRequirementIndex: 1,
              reportType: "General",
              reportDueDate: "2020-01-10T23:59:59.999-07:00",
              reportingRequirementId: 1,
              hasExpenses: true,
              calculatedGrossAmount: 567,
              calculatedNetAmount: 789,
              calculatedHoldbackAmount: 891,
              adjustedNetAmount: 89,
              adjustedGrossAmount: 67,
              adjustedHoldbackAmount: 91,
            },
          },
        },
      ],
    },
  };

  const mockQueryPayloadLatestCommitted = {
    ...mockQueryPayload,
    ProjectRevision() {
      const originalProjectRevision = mockQueryPayload.ProjectRevision();
      const modifiedProjectRevision = {
        ...originalProjectRevision,
        latestCommittedMilestoneFormChanges: {
          edges: [
            {
              node: {
                newFormData:
                  latestCommittedData.latestCommittedMilestoneFormChanges
                    .edges[0].node.newFormData,
              },
            },
          ],
        },
      };
      return modifiedProjectRevision;
    },
  };

  it("Displays diffs of the data fields that were updated and shows latest committed values", () => {
    componentTestingHelper.defaultQueryResolver =
      mockQueryPayloadLatestCommitted;
    componentTestingHelper.loadQuery(mockQueryPayloadLatestCommitted);
    componentTestingHelper.renderComponent();

    // calculated values
    expect(
      screen.getByText("Gross Payment Amount This Milestone")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$567\.00/i)).toBeInTheDocument();
    expect(screen.getByText(/\$999\.00/i)).toBeInTheDocument();
    expect(
      screen.getByText("Net Payment Amount This Milestone")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$789\.00/i)).toBeInTheDocument();
    expect(screen.getByText(/\$888\.00/i)).toBeInTheDocument();
    expect(
      screen.getByText("Holdback Amount This Milestone")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$891\.00/i)).toBeInTheDocument();
    expect(screen.getByText(/\$111\.00/i)).toBeInTheDocument();

    // adjusted values
    expect(
      screen.getByText("Gross Payment Amount This Milestone (Adjusted)")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$99\.00/i)).toBeInTheDocument();

    expect(screen.getByText(/\$89\.00/i)).toBeInTheDocument();
    expect(
      screen.getByText("Net Payment Amount This Milestone (Adjusted)")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$11\.00/i)).toBeInTheDocument();
    expect(screen.getByText(/\$67\.00/i)).toBeInTheDocument();
    expect(
      screen.getByText("Holdback Amount This Milestone (Adjusted)")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$88\.00/i)).toBeInTheDocument();
    expect(screen.getByText(/\$91\.00/i)).toBeInTheDocument();
  });
  const mockQueryPayloadWithDiffs = {
    Form() {
      return {
        jsonSchema: milestoneProdSchema,
      };
    },
    ProjectRevision() {
      const result = {
        isFirstRevision: false,
        summaryMilestoneFormChanges: {
          edges: [
            {
              node: {
                id: "Test Reporting Requirement ID - 1",
                isPristine: false,
                newFormData: {
                  totalEligibleExpenses: 1000,
                  description: "charmander",
                  projectId: 1,
                  reportingRequirementIndex: 1,
                  reportType: "General",
                  reportDueDate: "2020-01-10T23:59:59.999-07:00",
                  reportingRequirementId: 1,
                  hasExpenses: true,
                  calculatedNetAmount: 221,
                  calculatedGrossAmount: 222,
                  calculatedHoldbackAmount: 223,
                  adjustedNetAmount: 21,
                  adjustedGrossAmount: 22,
                  adjustedHoldbackAmount: 23,
                },
                operation: "UPDATE",
                formChangeByPreviousFormChangeId: {
                  newFormData: {
                    description: "bulbasaur",
                    projectId: 1,
                    reportingRequirementIndex: 1,
                    reportDueDate: "2020-01-01T13:59:59.999-07:00",
                    reportType: "Advanced",
                    reportingRequirementId: 1,
                    hasExpenses: true,
                    calculatedNetAmount: 111,
                    calculatedGrossAmount: 112,
                    calculatedHoldbackAmount: 113,
                    adjustedNetAmount: 11,
                    adjustedGrossAmount: 12,
                    adjustedHoldbackAmount: 13,
                  },
                },
                formDataRecordId: 1,
              },
            },
          ],
        },
      };
      return result;
    },
  };

  it("Displays diffs of the data fields that were updated and the old values", () => {
    componentTestingHelper.defaultQueryResolver = mockQueryPayloadWithDiffs;
    componentTestingHelper.loadQuery(mockQueryPayloadWithDiffs);
    componentTestingHelper.renderComponent();

    // calculated values
    expect(
      screen.getByText("Gross Payment Amount This Milestone")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$222\.00/i)).toBeInTheDocument();
    expect(screen.getByText(/\$112\.00/i)).toBeInTheDocument();
    expect(
      screen.getByText("Net Payment Amount This Milestone")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$221\.00/i)).toBeInTheDocument();
    expect(screen.getByText(/\$111\.00/i)).toBeInTheDocument();
    expect(
      screen.getByText("Holdback Amount This Milestone")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$223\.00/i)).toBeInTheDocument();
    expect(screen.getByText(/\$113\.00/i)).toBeInTheDocument();

    // adjusted values
    expect(
      screen.getByText("Gross Payment Amount This Milestone (Adjusted)")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$22\.00/i)).toBeInTheDocument();
    expect(screen.getByText(/\$12\.00/i)).toBeInTheDocument();
    expect(
      screen.getByText("Net Payment Amount This Milestone (Adjusted)")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$23\.00/i)).toBeInTheDocument();
    expect(screen.getByText(/\$13\.00/i)).toBeInTheDocument();
    expect(
      screen.getByText("Holdback Amount This Milestone (Adjusted)")
    ).toBeInTheDocument();
    expect(screen.getByText(/\$21\.00/i)).toBeInTheDocument();
    expect(screen.getByText(/\$11\.00/i)).toBeInTheDocument();
  });

  it("renders the tooltips for the mock summary", () => {
    componentTestingHelper.loadQuery();
    componentTestingHelper.renderComponent();

    expect(
      screen.getAllByRole("tooltip", {
        name: "maximum-amount-this-milestone-tooltip",
      })
    ).toHaveLength(2);

    expect(
      screen.getAllByRole("tooltip", {
        name: "total-eligible-expenses-tooltip",
      })
    ).toHaveLength(2);

    expect(
      screen.getAllByRole("tooltip", {
        name: "gross-payment-amount-this-milestone-tooltip",
      })
    ).toHaveLength(2);

    expect(
      screen.getAllByRole("tooltip", {
        name: "net-payment-amount-this-milestone-tooltip",
      })
    ).toHaveLength(2);

    expect(
      screen.getAllByRole("tooltip", {
        name: "holdback-amount-this-milestone-tooltip",
      })
    ).toHaveLength(2);
  });
});
