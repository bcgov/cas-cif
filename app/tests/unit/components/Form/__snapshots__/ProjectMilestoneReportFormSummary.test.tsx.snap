// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Project Milestone Report Form Summary renders the tooltips for the mock summary 1`] = `
<div>
  <h3>
    Project Milestone Reports
  </h3>
  <div
    class="jsx-1509299673 reportContainer"
  >
    <header
      class="jsx-1509299673"
    >
      <h4
        class="jsx-1509299673"
      >
        Milestone Report 
        1
      </h4>
    </header>
    <dl
      class="jsx-337939564 "
      novalidate=""
    >
      <div
        class="jsx-2146271977"
      >
        <div
          class="jsx-2146271977 definition-container"
        >
          <div>
            <div
              class="jsx-2146271977"
            >
              <div
                class="jsx-2146271977 definition-container"
              >
                <dt
                  class="jsx-156197991"
                >
                  Milestone Description 
                </dt>
                <dd
                  class="jsx-41823073"
                >
                  <em
                    class="jsx-41823073"
                  >
                    Not added
                  </em>
                </dd>
                <dd
                  class="jsx-41823073 contentSuffix"
                />
              </div>
            </div>
            <div
              class="jsx-2146271977"
            >
              <div
                class="jsx-2146271977 definition-container"
              >
                <dt
                  class="jsx-156197991"
                >
                  Milestone Type 
                </dt>
                <dd
                  class="jsx-41823073"
                >
                  General Milestone
                </dd>
                <dd
                  class="jsx-41823073 contentSuffix"
                />
              </div>
            </div>
            <div
              class="jsx-2146271977"
            >
              <div
                class="jsx-2146271977 definition-container"
              >
                <input
                  id="root_hasExpenses"
                  type="hidden"
                  value="true"
                />
              </div>
            </div>
            <div
              class="jsx-2146271977"
            >
              <div
                class="jsx-2146271977 definition-container"
              >
                <dt
                  class="jsx-156197991"
                >
                  Maximum Amount This Milestone 
                  <span
                    aria-label="maximum-amount-this-milestone-tooltip"
                    class="tooltip jsx-2050293192"
                    data-mui-internal-clone-element="true"
                    role="tooltip"
                    tabindex="0"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-info-circle fa-w-16 "
                      color="#1A5A96"
                      data-icon="info-circle"
                      data-prefix="fas"
                      focusable="false"
                      name="tooltip"
                      role="img"
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                </dt>
                <dd
                  class="jsx-2240123068"
                >
                  <span
                    class="decimal"
                    id="root_maximumAmount"
                  >
                    $23.00
                  </span>
                  <span
                    class="jsx-2240123068 contentSuffix"
                  />
                </dd>
              </div>
            </div>
            <div
              class="jsx-2146271977"
            >
              <div
                class="jsx-2146271977 definition-container"
              >
                <dt
                  class="jsx-156197991"
                >
                  Substantial Completion Date 
                </dt>
                <dd
                  class="jsx-41823073"
                >
                  <em
                    class="jsx-41823073"
                  >
                    Not added
                  </em>
                </dd>
                <dd
                  class="jsx-41823073 contentSuffix"
                />
              </div>
            </div>
            <div
              class="jsx-2146271977"
            >
              <div
                class="jsx-2146271977 definition-container"
              >
                <dt
                  class="jsx-156197991"
                >
                  Report Due Date 
                </dt>
                <dd
                  class="jsx-41823073"
                >
                  <em
                    class="jsx-41823073"
                  >
                    Not added
                  </em>
                </dd>
                <dd
                  class="jsx-41823073 contentSuffix"
                />
              </div>
            </div>
            <div
              class="jsx-2146271977"
            >
              <div
                class="jsx-2146271977 definition-container"
              >
                <dt
                  class="jsx-156197991"
                >
                  Certifier (optional) 
                </dt>
                <dd
                  class="jsx-41823073"
                >
                  <em
                    class="jsx-41823073"
                  >
                    Not added
                  </em>
                </dd>
                <dd
                  class="jsx-41823073 contentSuffix"
                />
              </div>
            </div>
            <div
              class="jsx-2146271977"
            >
              <div
                class="jsx-2146271977 definition-container"
              >
                <dt
                  class="jsx-156197991"
                >
                  Professional Designation (optional) 
                </dt>
                <dd
                  class="jsx-41823073"
                >
                  Professional Engineer
                </dd>
                <dd
                  class="jsx-41823073 contentSuffix"
                />
              </div>
            </div>
            <div
              class="jsx-2146271977"
            >
              <div
                class="jsx-2146271977 definition-container"
              >
                <dt
                  class="jsx-156197991"
                >
                  Report Received Date (optional) 
                </dt>
                <dd
                  class="jsx-41823073"
                >
                  <em
                    class="jsx-41823073"
                  >
                    Not added
                  </em>
                </dd>
                <dd
                  class="jsx-41823073 contentSuffix"
                />
              </div>
            </div>
            <div
              class="jsx-2146271977"
            >
              <div
                class="jsx-2146271977 definition-container"
              >
                <dt
                  class="jsx-156197991"
                >
                  Total Eligible Expenses (optional) 
                  <span
                    aria-label="total-eligible-expenses-tooltip"
                    class="tooltip jsx-2050293192"
                    data-mui-internal-clone-element="true"
                    role="tooltip"
                    tabindex="0"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-info-circle fa-w-16 "
                      color="#1A5A96"
                      data-icon="info-circle"
                      data-prefix="fas"
                      focusable="false"
                      name="tooltip"
                      role="img"
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                </dt>
                <dd
                  class="jsx-2240123068"
                >
                  <span
                    class="decimal"
                    id="root_totalEligibleExpenses"
                  >
                    $1,000.00
                  </span>
                  <span
                    class="jsx-2240123068 contentSuffix"
                  />
                </dd>
              </div>
            </div>
            <div
              class="jsx-2146271977"
            >
              <div
                class="jsx-2146271977 definition-container"
              >
                <dt
                  class="jsx-156197991"
                >
                  Gross Payment Amount This Milestone
                  <span
                    aria-label="gross-payment-amount-this-milestone-tooltip"
                    class="tooltip jsx-2050293192"
                    data-mui-internal-clone-element="true"
                    role="tooltip"
                    tabindex="0"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-info-circle fa-w-16 "
                      color="#1A5A96"
                      data-icon="info-circle"
                      data-prefix="fas"
                      focusable="false"
                      name="tooltip"
                      role="img"
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                </dt>
                <div
                  class="jsx-2935470220 readOnlyAdjustableCalculatedValueWidgetWrapper"
                >
                  <span
                    id="root_adjustedGrossAmount"
                  >
                    $123.00
                  </span>
                  <div
                    class="jsx-2935470220 adjustedValue"
                  >
                    <dt
                      class="jsx-2935470220"
                    >
                      Gross Payment Amount This Milestone
                       (Adjusted)
                    </dt>
                    <dd
                      class="jsx-2935470220"
                    >
                      <span
                        id="root_adjustedGrossAmount_adjusted"
                      >
                        $1,000.00
                      </span>
                    </dd>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="jsx-2146271977"
            >
              <div
                class="jsx-2146271977 definition-container"
              >
                <dt
                  class="jsx-156197991"
                >
                  Net Payment Amount This Milestone
                  <span
                    aria-label="net-payment-amount-this-milestone-tooltip"
                    class="tooltip jsx-2050293192"
                    data-mui-internal-clone-element="true"
                    role="tooltip"
                    tabindex="0"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-info-circle fa-w-16 "
                      color="#1A5A96"
                      data-icon="info-circle"
                      data-prefix="fas"
                      focusable="false"
                      name="tooltip"
                      role="img"
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                </dt>
                <div
                  class="jsx-2935470220 readOnlyAdjustableCalculatedValueWidgetWrapper"
                >
                  <span
                    id="root_adjustedNetAmount"
                  >
                    $789.00
                  </span>
                  <div
                    class="jsx-2935470220 adjustedValue"
                  >
                    <dt
                      class="jsx-2935470220"
                    >
                      Net Payment Amount This Milestone
                       (Adjusted)
                    </dt>
                    <em
                      class="jsx-2935470220"
                    >
                      Not added
                    </em>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="jsx-2146271977"
            >
              <div
                class="jsx-2146271977 definition-container"
              >
                <dt
                  class="jsx-156197991"
                >
                  Holdback Amount This Milestone
                  <span
                    aria-label="holdback-amount-this-milestone-tooltip"
                    class="tooltip jsx-2050293192"
                    data-mui-internal-clone-element="true"
                    role="tooltip"
                    tabindex="0"
                  >
                    <svg
                      aria-hidden="true"
                      class="svg-inline--fa fa-info-circle fa-w-16 "
                      color="#1A5A96"
                      data-icon="info-circle"
                      data-prefix="fas"
                      focusable="false"
                      name="tooltip"
                      role="img"
                      viewBox="0 0 512 512"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
                        fill="currentColor"
                      />
                    </svg>
                  </span>
                </dt>
                <div
                  class="jsx-2935470220 readOnlyAdjustableCalculatedValueWidgetWrapper"
                >
                  <span
                    id="root_adjustedHoldbackAmount"
                  >
                    $456.00
                  </span>
                  <div
                    class="jsx-2935470220 adjustedValue"
                  >
                    <dt
                      class="jsx-2935470220"
                    >
                      Holdback Amount This Milestone
                       (Adjusted)
                    </dt>
                    <em
                      class="jsx-2935470220"
                    >
                      Not added
                    </em>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="jsx-2146271977"
            >
              <div
                class="jsx-2146271977 definition-container"
              >
                <dt
                  class="jsx-156197991"
                >
                  Date Invoice Sent to CSNR (optional) 
                </dt>
                <dd
                  class="jsx-41823073"
                >
                  <em
                    class="jsx-41823073"
                  >
                    Not added
                  </em>
                </dd>
                <dd
                  class="jsx-41823073 contentSuffix"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </dl>
  </div>
</div>
`;
